### PR TITLE
th-212: Shift service

### PR DESCRIPTION
**Pull request by TetianaDoroshko**
There are following endpoints:
 GET shifts/  - for getting all opened shifts at the moment (open for everyone)
 POST shifts/  - for openning (create) a new shift  (available for business and driver)
 PUT shift/:id  - for closing (insert end date and soft delete) the shift (available for business and driver)
 GET shift/:id  - for getting the particular shift by its id  (open for everyone)
 GET shift/driver/:id  - for getting all shifts belonged to a the driver  (open for everyone)
 
 PS some methods aren't used and were created only for corresponding to core interface structures.

### CODE REVIEW

File: backend/src/packages/shifts/shift.repository.ts [hunk 1]
```diff
@@ -0,0 +1,109 @@
+import { and, desc, eq, isNull } from 'drizzle-orm';
+
+import { type IRepository } from '~/libs/interfaces/repository.interface.js';
+import { type IDatabase } from '~/libs/packages/database/libs/interfaces/database.interface.js';
```
Comment (Druxight) [inline backend/src/packages/shifts/shift.repository.ts]:
```suggestion
import { type IRepository } from '~/libs/interfaces/interfaces.js';
import { type IDatabase } from '~/libs/packages/database/libs/interfaces/interfaces.js';
```

File: backend/src/packages/shifts/shift.service.ts [hunk 2]
```diff
@@ -0,0 +1,132 @@
+import { type ShiftCloseRequestDto } from 'shared/build/index.js';
+
+import { type IService } from '~/libs/interfaces/service.interface';
```
Comment (Druxight) [inline backend/src/packages/shifts/shift.service.ts]:
```suggestion
import { type IService } from '~/libs/interfaces/interfaces.js';
```

File: backend/src/packages/shifts/shift.service.ts [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+
+  public async findById(id: number): Promise<ShiftCreateResponseDto | null> {
+    const [shift = null] = await this.shiftRepository.find({ id });
+
+    return shift ? ShiftEntity.initialize(shift).toObject() : null;
+  }
+
+  public async findByShiftId(
+    shiftId: number,
+    userId: number,
```
Comment (Druxight) [inline backend/src/packages/shifts/shift.service.ts]: I don't think we need the userId here. Correct me if I wrong

File: backend/src/packages/shifts/shift.service.ts [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+    const [shift = null] = await this.shiftRepository.find({ id: shiftId });
+
+    if (!shift) {
+      throw new HttpError({
+        status: HttpCode.NOT_FOUND,
+        message: HttpMessage.NOT_FOUND,
+      });
+    }
+
+    if (shift.driverUserId !== userId) {
```
Comment (Druxight) [inline backend/src/packages/shifts/shift.service.ts]: So our business is unable to get shifts data ?

File: backend/src/packages/shifts/shift.service.ts [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+        status: HttpCode.BAD_REQUEST,
+        message: HttpMessage.SHIFT_ALREADY_CLOSED,
+      });
+    }
+
+    const updShift = await this.shiftRepository.update({
+      id: params.id,
+      payload: {
+        ...body,
+        deletedAt: new Date(),
```
Comment (Druxight) [inline backend/src/packages/shifts/shift.service.ts]: please make sure that records with initialized deleted at are not returned by GET requests

File: backend/src/packages/shifts/shift.service.ts [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+    const shift = await this.findByShiftId(params.id, user.id);
+
+    if (shift.endDate !== null) {
+      throw new HttpError({
+        status: HttpCode.BAD_REQUEST,
+        message: HttpMessage.SHIFT_ALREADY_CLOSED,
+      });
+    }
+
+    const updShift = await this.shiftRepository.update({
```
Comment (Druxight) [inline backend/src/packages/shifts/shift.service.ts:123]: I think we should create a separate function on the repository layer called `close`

**Review by Druxight — CHANGES_REQUESTED** [review]

File: backend/src/packages/shifts/shift.service.ts (hunk 5 above)
Comment (TetianaDoroshko) [reply backend/src/packages/shifts/shift.service.ts]: This is PUT request. Has it to be DELETE?

**Comment by TetianaDoroshko** [comment]
Business user can open and close a shift only to a driver belonged to his busines.
Driver can open and close a shift only for himself.

File: backend/src/libs/packages/database/schema/tables-schema.ts [hunk 7]
```diff
@@ -119,6 +119,24 @@ const usersTrucks = pgTable(
...
 
+const shifts = pgTable('shifts', {
+  id: serial('id').primaryKey(),
+  startDate: timestamp('start_date', {
+    mode: 'date',
+  }).notNull(),
+  endDate: timestamp('end_date', {
+    mode: 'date',
+  }),
+  driverUserId: integer('driver_user_id')
```
Comment (Druxight) [inline backend/src/libs/packages/database/schema/tables-schema.ts]: rename to the `driverId`

**Review by Druxight — CHANGES_REQUESTED** [review]

**Review by Druxight — DISMISSED** [review]

**Review by DmytroVoronkov — APPROVED** [review]

File: shared/src/packages/shifts/libs/enums/shift-api-path.enum.ts [hunk 8]
```diff
@@ -0,0 +1,7 @@
+const ShiftsApiPath = {
+  ROOT: '/',
+  $ID: '/:id',
+  $DRIVER_ID: '/driver',
```
Comment (kalinkaaaa14) [inline shared/src/packages/shifts/libs/enums/shift-api-path.enum.ts]:
```suggestion
  DRIVER_$ID: '/driver/:driverId',
```

File: backend/src/packages/shifts/shift.repository.ts [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+    id: ShiftEntityT['id'],
+  ): Promise<ShiftDatabaseModel | null> {
+    const [shift] = await this.db
+      .driver()
+      .select()
+      .from(this.shiftSchema)
+      .where(eq(this.shiftSchema.id, id))
+      .execute();
+
+    return shift;
```
Comment (kalinkaaaa14) [inline backend/src/packages/shifts/shift.repository.ts:76]: shift ?? null maybe?

File: backend/src/packages/shifts/shift.service.ts [hunk 10]
```diff
@@ -0,0 +1,184 @@
...
+    } else if (user.group.key === UserGroupKey.DRIVER) {
+      if (user.id !== driverId) {
+        return false;
+      }
+    } else {
+      return false;
+    }
+
+    return true;
+  }
```
Comment (kalinkaaaa14) [inline backend/src/packages/shifts/shift.service.ts]:
```suggestion
 
    const { group } = user;
    const { key: groupKey } = group;

    if (groupKey === UserGroupKey.BUSINESS) {
     const isDriverBelongedToBusiness = await checkisDriverBelongedToBusiness({businessId, driverId});

     return isDriverBelongedToBusiness;
    }

    if(groupKey === UserGroupKey.DRIVER) {
      return user.id === driverId;
    }

    return false;
```

Comment (kalinkaaaa14) [reply backend/src/packages/shifts/shift.service.ts]: and naming - checkHasAccessToShift

File: shared/src/packages/shifts/libs/types/shift-entity.type.ts [hunk 11]
```diff
@@ -0,0 +1,13 @@
+import { type UserEntityT } from '~/packages/users/users.js';
+
+import { type TruckEntity } from '../../../../packages/trucks/trucks.js';
+
+type ShiftEntity = {
```
Comment (kalinkaaaa14) [inline shared/src/packages/shifts/libs/types/shift-entity.type.ts:4]: maybe change naming, entities are mostly in be folder
better type ShiftResponseDto, and then you don't need to create a type ShiftCreateResponseDto

**Review by kalinkaaaa14 — CHANGES_REQUESTED** [review]

File: backend/src/packages/shifts/shift.service.ts (hunk 10 above)
Comment (kalinkaaaa14) [reply backend/src/packages/shifts/shift.service.ts]: and can we make this helper checkisDriverBelongedToBusiness - with only one db call?

Comment (TetianaDoroshko) [reply backend/src/packages/shifts/shift.service.ts]: I've refactored the method and the db call

File: shared/src/packages/shifts/libs/types/shift-entity.type.ts (hunk 11 above)
Comment (TetianaDoroshko) [reply shared/src/packages/shifts/libs/types/shift-entity.type.ts:4]: if you allow, I'll stay this type, it is convenient for understanding of entity and request-response structure for me. I've only renamed ShiftCreateResponseDto to ShiftResponseDto.

File: shared/src/packages/shifts/libs/types/shift-entity.type.ts [hunk 12]
```diff
@@ -0,0 +1,13 @@
+import { type UserEntityT } from '~/packages/users/users.js';
+
+import { type TruckEntity } from '../../../../packages/trucks/trucks.js';
```
Comment (kalinkaaaa14) [inline shared/src/packages/shifts/libs/types/shift-entity.type.ts]:
```suggestion
import { type TruckEntity } from '~/packages/trucks/trucks.js';
```

File: shared/src/packages/shifts/libs/enums/shift-api-path.enum.ts [hunk 13]
```diff
@@ -0,0 +1,7 @@
+const ShiftsApiPath = {
+  ROOT: '/',
+  $ID: '/:id',
+  DRIVER_$ID: '/driver',
```
Comment (kalinkaaaa14) [inline shared/src/packages/shifts/libs/enums/shift-api-path.enum.ts]:
```suggestion
  DRIVER_$ID: '/driver/:driverId',
```
I don't like concatenation of strings for  building a path how it's now, replace it

File: backend/src/packages/shifts/shift.controller.ts [hunk 14]
```diff
@@ -0,0 +1,153 @@
...
+
+class ShiftController extends Controller {
+  private shiftService: ShiftService;
+
+  public constructor(logger: ILogger, shiftService: ShiftService) {
+    super(logger, ApiPath.SHIFTS);
+
+    this.shiftService = shiftService;
+
+    this.addRoute({
```
Comment (kalinkaaaa14) [inline backend/src/packages/shifts/shift.controller.ts:34]: add swagger doc

**Review by kalinkaaaa14 — DISMISSED** [review]

File: backend/src/packages/shifts/shift.service.ts [hunk 15]
```diff
@@ -0,0 +1,182 @@
...
+  public async findByDriverUserId(
+    driverId: number,
+  ): Promise<ShiftResponseDto[]> {
+    const shifts = await this.shiftRepository.find({
+      driverId,
+    });
+
+    return shifts.length > 0
+      ? shifts.map((it) => ShiftEntity.initialize(it).toObject())
+      : [];
```
Comment (PavloSukhinin) [inline backend/src/packages/shifts/shift.service.ts]: same as above

File: backend/src/packages/shifts/shift.service.ts [hunk 16]
```diff
@@ -0,0 +1,182 @@
...
+    this.businessService = businessService;
+    this.driverService = driverService;
+  }
+
+  public async getAllStarted(): Promise<ShiftResponseDto[]> {
+    const shifts = await this.shiftRepository.getAllOpened();
+
+    return shifts.length > 0
+      ? shifts.map((it) => ShiftEntity.initialize(it).toObject())
+      : [];
```
Comment (PavloSukhinin) [inline backend/src/packages/shifts/shift.service.ts]:
```suggestion
    return shifts.map((it) => ShiftEntity.initialize(it).toObject())
```

[].map() returns []

File: shared/src/packages/shifts/libs/types/shift-close-request-dto.type.ts [hunk 17]
```diff
@@ -0,0 +1,3 @@
+type ShiftCloseRequestDto = { endDate: Date };
```
Comment (PavloSukhinin) [inline shared/src/packages/shifts/libs/types/shift-close-request-dto.type.ts:1]: Just note: we can't pass Date instance through network

**Review by PavloSukhinin — CHANGES_REQUESTED** [review]
Don't forget to update schema in readme.md

File: shared/src/packages/shifts/libs/types/shift-close-request-dto.type.ts (hunk 17 above)
Comment (TetianaDoroshko) [reply shared/src/packages/shifts/libs/types/shift-close-request-dto.type.ts:1]: yes, it actually comes in string (formatted like date). Do I need to replace type to string and validatate it by pattern (like /\d{4}-[01]\d-[0-3]\dT[0-2]\d:[0-5]\d:[0-5]\d\.\d+([+-][0-2]\d:[0-5]\d|Z)/ ) ?

Comment (PavloSukhinin) [reply shared/src/packages/shifts/libs/types/shift-close-request-dto.type.ts:1]: Let's keep it as is for now

**Review by PavloSukhinin — APPROVED** [review]

**Pull request merged**